app/src/kscan_sideband_behaviors.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2023 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_kscan_sideband_behaviors
+
+#include <zephyr/device.h>
+#include <drivers/behavior.h>
+#include <zephyr/drivers/kscan.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+
+#include <zmk/event_manager.h>
+#include <zmk/behavior.h>
+#include <zmk/keymap.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+struct ksbb_entry {
+    uint8_t row;
+    uint8_t column;
+    struct zmk_behavior_binding binding;
+};
+
+struct ksbb_config {
+    const struct device *kscan;
+    struct ksbb_entry *entries;
+    size_t entries_len;
+};
+
+struct ksbb_data {
+    kscan_callback_t callback;
+    bool enabled;
+};
+
+#define GET_KSBB_DEV(n) DEVICE_DT_GET(DT_DRV_INST(n)),
+
+// The kscan callback has no context with it, so we keep a static array of all possible
+// KSBBs to check when a kscan callback from the "wrapped" inner kscan fires.
+static const struct device *ksbbs[DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)] = {
+    DT_INST_FOREACH_STATUS_OKAY(GET_KSBB_DEV)};
+
+void ksbb_inner_kscan_callback(const struct device *dev, uint32_t row, uint32_t column,
+                               bool pressed) {
+    for (int i = 0; i < ARRAY_SIZE(ksbbs); i++) {
+        const struct device *ksbb = ksbbs[i];
+        const struct ksbb_config *cfg = ksbb->config;
+        struct ksbb_data *data = ksbb->data;
+
+        if (cfg->kscan != dev) {
+            continue;
+        }
+
+        for (int e = 0; e < cfg->entries_len; e++) {
+            struct ksbb_entry *entry = &cfg->entries[e];
+            if (entry->row == row && entry->column == column) {
+                struct zmk_behavior_binding_event event = {.position = INT32_MAX,
+                                                           .timestamp = k_uptime_get()};
+
+                if (pressed) {
+                    behavior_keymap_binding_pressed(&entry->binding, event);
+                } else {
+                    behavior_keymap_binding_released(&entry->binding, event);
+                }
+                return;
+            }
+        }
+
+        if (data->enabled && data->callback) {
+            data->callback(ksbb, row, column, pressed);
+        }
+    }
+}
+
+static int ksbb_configure(const struct device *dev, kscan_callback_t callback) {
+    const struct ksbb_config *cfg = dev->config;
+    struct ksbb_data *data = dev->data;
+
+    if (!callback) {
```

Comment:
Based on the check on line 71, callback being NULL appears to be a valid state, so I assume it is okay if you don't set a callback to begin with. Is it not allowed to set the callback back to NULL after previously setting it to something?

Reply:
Yeah, this isn't really necessary. Removed.